src/main/groovy/com/pszymczyk/consul/infrastructure/SimpleConsulClient.groovy

```diff
@@ -1,17 +1,36 @@
 package com.pszymczyk.consul.infrastructure
 
 import groovyx.net.http.ContentType
+import groovyx.net.http.EncoderRegistry
 import groovyx.net.http.HttpResponseDecorator
 import groovyx.net.http.RESTClient
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
 
 class SimpleConsulClient {
 
-    private static final String NO_LEADER_ELECTED_RESPONSE = "";
+    private static final String NO_LEADER_ELECTED_RESPONSE = ""
 
     private final RESTClient http
+    private final String token
 
-    SimpleConsulClient(String host, int httpPort) {
+    SimpleConsulClient(String host, int httpPort, String token) {
         this.http = new RESTClient("http://$host:$httpPort")
+
+        this.token = token
+        if (token != null) {
+            this.http.setHeaders('X-Consul-Token': token)
+        }
+
+        def originalParser = this.http.parser.getAt(ContentType.JSON)
+        this.http.parser.putAt(ContentType.JSON, { HttpResponseDecorator resp ->
+            if (resp.status >= 200 && resp.status < 300 ){
+                return originalParser(resp)
+            } else {
+                throw new GroovyRuntimeException("Unexpected response status code: ${resp.status}:" +
```

Comment:
Why do we need this parser? It is out of scope change IMHO

Reply:
Test `unable to get registered nodes with wrong token` is green, but prints tons of stacktraces:
```
The current character read is 'A' with an int value of 65
Unable to determine the current character, it is not a string, number, array, or object
line number 1
index number 0
ACL not found                                                                                                                                                                                                                                                   
^
	at groovy.json.internal.JsonParserCharArray.decodeValueInternal(JsonParserCharArray.java:206)
	at groovy.json.internal.JsonParserCharArray.decodeValue(JsonParserCharArray.java:157)
	at groovy.json.internal.JsonParserCharArray.decodeFromChars(JsonParserCharArray.java:46)
	at groovy.json.internal.JsonParserCharArray.parse(JsonParserCharArray.java:384)
	at groovy.json.internal.BaseJsonParser.parse(BaseJsonParser.java:128)
	at groovy.json.JsonSlurper.parse(JsonSlurper.java:221)
	at groovyx.net.http.ParserRegistry.parseJSON(ParserRegistry.java:280)
	at sun.reflect.GeneratedMethodAccessor9.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1215)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1082)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1024)
	at groovy.lang.Closure.call(Closure.java:414)
	at groovy.lang.Closure.call(Closure.java:430)
	at groovyx.net.http.HTTPBuilder.parseResponse(HTTPBuilder.java:560)
	at groovyx.net.http.HTTPBuilder$1.handleResponse(HTTPBuilder.java:489)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:223)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:165)
	at groovyx.net.http.HTTPBuilder.doRequest(HTTPBuilder.java:515)
	at groovyx.net.http.RESTClient.get(RESTClient.java:119)
	at groovyx.net.http.RESTClient$get.call(Unknown Source)
	at com.pszymczyk.consul.infrastructure.SimpleConsulClient.getRegisteredNodes(SimpleConsulClient.groovy:43)
	at com.pszymczyk.consul.infrastructure.SimpleConsulClient$getRegisteredNodes$0.call(Unknown Source)
	at com.pszymczyk.consul.infrastructure.ConsulWaiter.allNodesRegistered(ConsulWaiter.groovy:75)
	at sun.reflect.GeneratedMethodAccessor12.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:210)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:59)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:158)
	at com.pszymczyk.consul.infrastructure.ConsulWaiter.awaitUntilConsulStarted(ConsulWaiter.groovy:41)
	at com.pszymczyk.consul.infrastructure.ConsulWaiter$awaitUntilConsulStarted.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:117)
	at com.pszymczyk.consul.ConsulStarter.start(ConsulStarter.groovy:153)
	at com.pszymczyk.consul.ConsulStarter$start.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:117)
	at com.pszymczyk.consul.infrastructure.ConsulWaiterTest.$spock_feature_0_2(ConsulWaiterTest.groovy:64)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.spockframework.util.ReflectionUtil.invokeMethod(ReflectionUtil.java:153)
	at org.spockframework.runtime.model.MethodInfo.invoke(MethodInfo.java:84)
	at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:480)
	at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:463)
	at org.spockframework.runtime.BaseSpecRunner.runFeatureMethod(BaseSpecRunner.java:405)
	at org.spockframework.runtime.BaseSpecRunner.doRunIteration(BaseSpecRunner.java:324)
	at org.spockframework.runtime.BaseSpecRunner$6.invoke(BaseSpecRunner.java:309)
	at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:480)
	at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:463)
	at org.spockframework.runtime.BaseSpecRunner.runIteration(BaseSpecRunner.java:288)
	at org.spockframework.runtime.BaseSpecRunner.initializeAndRunIteration(BaseSpecRunner.java:278)
	at org.spockframework.runtime.BaseSpecRunner.runSimpleFeature(BaseSpecRunner.java:269)
	at org.spockframework.runtime.BaseSpecRunner.doRunFeature(BaseSpecRunner.java:263)
	at org.spockframework.runtime.BaseSpecRunner$5.invoke(BaseSpecRunner.java:246)
	at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:480)
	at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:463)
	at org.spockframework.runtime.BaseSpecRunner.runFeature(BaseSpecRunner.java:238)
	at org.spockframework.runtime.BaseSpecRunner.runFeatures(BaseSpecRunner.java:188)
	at org.spockframework.runtime.BaseSpecRunner.doRunSpec(BaseSpecRunner.java:98)
	at org.spockframework.runtime.BaseSpecRunner$1.invoke(BaseSpecRunner.java:84)
	at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:480)
	at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:463)
	at org.spockframework.runtime.BaseSpecRunner.runSpec(BaseSpecRunner.java:76)
	at org.spockframework.runtime.BaseSpecRunner.run(BaseSpecRunner.java:67)
	at org.spockframework.runtime.Sputnik.run(Sputnik.java:63)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

If I am the client, I do not want to dive into this chaos. Do you think it is reasonable?

Or are you saying about adding more concrete asserts?

Reply:
As I can see case with incorrect token is not connected with JSON parsing, under the hood http client throws `groovyx.net.http.ResponseParseException: Forbidden` and `403` status code which is correct behaviour but groovy translate it to `org.codehaus.groovy.runtime.InvokerInvocationException` which is wired. I will check it in coming days and get back with solution